src/transformers/singleton-unions.ts

```diff
@@ -0,0 +1,37 @@
+import { JSONSchema4 } from 'json-schema';
+
+/**
+ * Schemas may define a type as a union of types with one of the types being `null`.
+ * This is the same as defining an optional type.
+ */
+export function hoistSingletonUnions(def: JSONSchema4): JSONSchema4 {
+  // definitions already has a type, can't hoist up
+  if (def.type) {
+    return def;
+  }
+
+  // if we have more than one of anyOf, oneOf, or allOf, behavior is undefined and we can't hoist up
+  const hasAnyOf = Boolean(def.anyOf);
+  const hasOneOf = Boolean(def.oneOf);
+  const hasAllOf = Boolean(def.allOf);
+  if ([hasAnyOf, hasOneOf, hasAllOf].filter(value => value === true).length > 1) {
```

Comment:
Isn't this the same as:

```suggestion
  if (hasAnyOf || hasOneOf || hasAllOf) {
```